chingu-voyages/v51-tier3-team-33
# Gary/feature/update delete expenses

This builds on top of my 2 other PRs. Please review them first before this one.

# Description
This commit implements the ability to update and delete existing expenses. 

## Related to the following Jira/Trello issue 

I will link it if I can find one after.

## Main changes explained:
- Created API to handle updating existing expenses.
- Created API to handle deletion of existing expenses.
- Created deleteFromS3 method in S3-Upload file to handle deletion of existing expense receipts.
- Moved S3-Upload file to Groups folder instead of its own thing.
- Added feature to allow contributions to be added and updated in addExpense and updateExpense apis.
- Added ability to update an expense as "paid".

## How to test:

Since this branch includes 2 main features.  Here are two different sets of instructions. I 
0. I'm sorry in advance. 
1. check into current branch
2. do `pnpm install` and `pnpm run dev` to run this PR locally
3. Make sure you are logged in.

NOTE: I added contributions to this branch. To test this add contributions as a key in the form data, the value should be an array of objects that will have this format. Ex: [{"member_id": "66f83db2e64fda3c2783ae10", "amount": 100}, {"member_id": "66f83e89dcdaaa757d8642b2", "amount": 50]}. 

I also added an is_paid field, this is just a simple way to allow an expense be updated as completed. the key for this in the form data is "is_paid" the value will need to be set to either true, or false.

Updating Expenses:
1. Create a new group.
2. Add at least 2 - 3 users to the newly created group via a post request.
3. Add an expense to the new group the URL of the request should look like localhost:3000/api/groups/ *insert group id here*/ addExpense - Make sure you are entering the expense data as form-data using key-value pairs, otherwise you will get an error.
5. Update the newly created expense, the URL of the request should look like localhost:3000/api/groups/*insert group id here*/expenses/*insert expense id here*/updateExpense. You want to make sure all fields are present with some of the values changed however you want. Verify the expense returned matches the new changes.
  
Deleting Expenses:
1. Delete the updated expense, the URL of the request should look like localhost:3000/api/groups/*insert group id here*/expenses/*insert expense id here*/deleteExpense. 
2. Check the response, and verify the message says that the expense is deleted.
3. Confirm that the expense is deleted by making sure that no UserExpenses have the Id of the expense you created. 
5. Confirm that the expense is deleted by making sure that the expense is no longer shown in the expenses array of the group.
6. Confirm that the expense is deleted by making that the expense is no longer present in the list of expenses.



## Note:
I thought this would be much less complex, hence why I had these both be on one branch. Sorry about that.
I also know the url of the requests can be quite long. If its a problem, I can try to shorten them, Just let me know!
